src/command-source.c

```diff
@@ -156,39 +173,161 @@ command_source_get_property (GObject      *object,
-    if (new_fd > command_source->maxfd)
-        command_source->maxfd = new_fd;
-    g_debug ("command_source_on_new_connection: writing \"%s\" to fd: %d",
-             WAKEUP_DATA, command_source->wakeup_send_fd);
-    ret = write (command_source->wakeup_send_fd, WAKEUP_DATA, WAKEUP_SIZE);
-    if (ret < 1)
-        g_error ("failed to send wakeup signal");
+    g_info ("%s: adding new connection: 0x%" PRIxPTR, __func__, (uintptr_t)connection);
+    /*
+     * Take reference to socket, will be freed when the source_data_t
+     * structure is freed
+     */
+    g_object_ref (socket);
+    data = g_malloc0 (sizeof (source_data_t));
+    data->cancellable = g_cancellable_new ();
+    data->source = g_socket_create_source (socket, G_IO_IN, data->cancellable);
+    /* we ignore the ID returned since we keep a reference to the source around */
+    g_source_attach (data->source, self->main_context);
+    data->self = self;
+    g_source_set_callback (data->source,
+                           (GSourceFunc)command_source_on_io_ready,
+                           data,
+                           NULL);
+    /*
+     * To stop watching this socket for G_IO_IN condition use this GHashTable
+     * to look up the GCancellable object. The hash table takes ownership of
+     * the reference to the socket and the source_data_t pointer.
+     */
+    g_hash_table_insert (self->socket_to_source_data_map, socket, data);
 
     return 0;
 }
+/*
+ * callback for iterating over objects in the member socket_to_source_data_map
+ * GHashMap to kill off the GSource. This requires cancelling it, and then
+ * calling the GSource destroy function. This should only happen when the
+ * CommandSource object is destroyed.
+ */
+static void
+command_source_source_cancel (gpointer key,
+                              gpointer value,
+                              gpointer user_data)
+{
+    g_debug ("%s", __func__);
+    source_data_t *data = (source_data_t*)value;
+
+    g_debug ("%s: canceling cancellable: 0x%" PRIxPTR " and destroying "
+             "source: 0x%" PRIxPTR, __func__, (uintptr_t)data->cancellable,
+             (uintptr_t)data->source);
+    g_cancellable_cancel (data->cancellable);
+}
+/*
+ * GObject dispose function. It's used to unref / release all GObjects held
+ * by the CommandSource.
+ */
+static void
+command_source_dispose (GObject *object) {
+    CommandSource *self = COMMAND_SOURCE (object);
+
+    g_debug ("%s", __func__);
+    /* cancel all outstanding G_IO_IN conndition GSources and destroy them */
+    g_hash_table_foreach (self->socket_to_source_data_map,
+                          command_source_source_cancel,
+                          NULL);
+    g_debug ("%s: clearing socket_to_source_data_map with g_hash_table_unref",
+             __func__);
+    g_clear_pointer (&self->socket_to_source_data_map, g_hash_table_unref);
+    g_debug ("%s: done", __func__);
+}
 
 /* Not doing any sanity checks here. Be sure to shut things downon your own
```

Comment:
miss a space between down and on.

Reply:
good catch!

Reply:
oh wow that's an old comment, not from this patch. It's not accurate any more. I'll fix this up in a separate commit.